keysync_initiator.go

```diff
@@ -0,0 +1,172 @@
+package nitriding
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hf/nitrite"
+	"golang.org/x/crypto/nacl/secretbox"
+)
+
+// RequestKeys asks a remote enclave to share its key material with us, which
+// is then written to the provided variable.
+//
+// This is only necessary if you intend to scale enclaves using Kubernetes.  If
+// you will only ever run a single enclave, ignore this function.
+func RequestKeys(addr string, keyMaterial any) error {
+	errStr := "failed to request key material"
+
+	// First, request a nonce from the remote enclave.
+	theirNonce, err := requestNonce(addr)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Now, create our own nonce.
+	ourNonce, err := newNonce()
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Next, create a key that the remote enclave is going to use to encrypt
+	// its key material.
+	sbKey, err := newSbKey()
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Now create an attestation document containing our nonce, the remote
+	// enclave's nonce, and the key material that they remote enclave is
+	// supposed to use.
+	ourAttDoc, err := attest(theirNonce[:], ourNonce[:], sbKey.Bytes())
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Send our attestation document to the remote enclave, and get theirs in
+	// return.
+	theirAttDoc, err := requestAttDoc(addr, ourAttDoc)
+	if err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	// Finally, verify the attestation document and extract the key material.
+	if err := processAttDoc(theirAttDoc, &ourNonce, sbKey, keyMaterial); err != nil {
+		return fmt.Errorf("%s: %s", errStr, err)
+	}
+
+	return nil
+}
+
+// requestNonce requests a nonce from the remote enclave specified by 'addr'.
+func requestNonce(addr string) (nonce, error) {
+	errStr := "failed to fetch nonce from remote enclave"
+
+	endpoint := fmt.Sprintf("%s%s", addr, pathNonce)
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return nonce{}, fmt.Errorf("%s: %s", errStr, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
```

Comment:
What about limiting this to 3/2*nonceLen bytes?

Reply:
Fixed in 598600f.